### ISSUE TITLE
set backend attribute to indicate pdf 

### ISSUE THREAD

**Issue by danyill** [Mogztter/asciidoctor-web-pdf#129]
Currently the value of the `{backend}` attribute is html5.

I propose it be either `pdf` or `html` or `html-pdf` or `html5-pdf`. 

Although the toolchain involves the html, the end user is targetting pdf and may have logic (as I do) which seeks to deactivate features or modify the document in response to that.

**Comment by ahus1** [comment]
Currently the "backend" is the converter from AsciiDoc to HTML, and you could use different converters, for example [html5s, the semantic HTML converter](https://github.com/jirutka/asciidoctor-html5s).

For now I recommend that you add your own attribute on the command line/build process like this: 

`asciidoctor-pdf -a testattr=abc basic-example.adoc`

Looking forward (and to be discussed) the appropriate attribute I'd see is `{env}`. You can use it or `{env-github}` today to find out if your source [is rendered for example on GitHub for preview](https://gist.github.com/dcode/0cfbf2699a1fe9b46ff04c41721dda74#github-specific-customizations):

```asciidoc
ifdef::env-github[]
:imagesdir: foo/
endif::[]
```

I wonder if this this should be `pdfjs` or `pdf`.

**Comment by Mogztter** [comment]
I was really confused with the backend and converter terminology:

> A converter is the software component in Asciidoctor that performs conversion from AsciiDoc to a publishable format. A backend is an alias—​kind of like a keyword—​that identifies the converter to use.
The backend expresses an intent to transform to a certain format (e.g., html5 for HTML5). But it’s more than just being shorter to type. Multiple converters can bind to (i.e., stake claim to) the same backend in order to provide different approaches for performing that conversion. For example, the backend pdf could activate Asciidoctor PDF, but it’s also possible it means a different implementation. The last converter bound to a backend wins.

https://asciidoctor.org/docs/faq/#converter-vs-backend

So if we follow this principle we should use `pdf`.
I'm a little reluctant to do that because some users are using the backend in conditional directives. 

> Currently the "backend" is the converter from AsciiDoc to HTML, and you could use different converters, for example html5s, the semantic HTML converter.

Yes, even though, I think that the backend of the semantic HTML converter should be `html5` (not `html5s`).

> Looking forward (and to be discussed) the appropriate attribute I'd see is {env}. You can use it or {env-github} today to find out if your source is rendered for example on GitHub for preview:

I agree.
Ideally you shouldn't use the backend in a conditional block. I wonder if the converter name should be available via an AsciiDoc attribute...
The reveal.js converter is using the following:

- backend: `revealjs` (should be `reveal.js`)
- basebackend: `html`

> I wonder if this this should be pdfjs or pdf.

I think `js` suffix/prefix is not relevant, we should use either `html` or `web` to be consistent with the future new name of this project: https://github.com/Mogztter/asciidoctor-pdf.js/issues/125#issuecomment-555103151

```
ifdef::env-web-pdf[]
:imagesdir: foo/
endif::[]
```

In summary:

- `backend=pdf`
- `basebackend=html` (not sure about this one)
- `outfilesuffix=.pdf`
- `filetype=pdf`
- `env-web-pdf`
- `env=web-pdf`

@mojavelinux Thoughts?

**Comment by ahus1** [comment]
For me `backend` is what I pass on to the command line of Asciidoctor using `-b` command line option. 

Therefore I would keep `html5` (or `html5s` if someone chooses a different converter) for backend. 

I'm ok with `outfilesuffix`, `filetype`, `env-web-pdf` and `env`. I can't think of a situation where I would use `basebackend`

**Comment by mojavelinux** [comment]
I'll admit, this one is tricky.

The backend is strictly intended for converter resolution.

The idea of basebackend is to let you know what format inline content is in (for extensions). Although it's a little sketchy, Asciidoctor PDF uses pseudo-HTML for inline content, which is why it sets basebackend to html.

Since this conversion relies on the HTML converter to run first, I'm inclined to say the backend should remain unchanged or be html-pdf. You could also consider introducing a new attribute to give an indication that pdf is the intended output format.

**Comment by Mogztter** [comment]
> The backend is strictly intended for converter resolution.

Ok.

> Since this conversion relies on the HTML converter to run first, I'm inclined to say the backend should remain unchanged or be html-pdf.

At first, I thought that the conversion to HTML was just an implementation detail. An intermediate step to produce a PDF. But now I don't think it's true. The conversion to HTML is a central piece of this converter.

Why not `web-pdf`? I think it's important to be consistent in the naming, so if we rename this project `asciidoctor-web-pdf` then I think the backend should be `web-pdf`.
I also think that `html-pdf` can be confusing because both `html` and `pdf` are output formats.

> You could also consider introducing a new attribute to give an indication that pdf is the intended output format.

`outfileformat=pdf`? `outfilemimetype=application/pdf`? I don't use separators because `outfilesuffix` does not use separators but arguably it's hard to read...

What do you think about `env-web-pdf` and `env=web-pdf`?

**Comment by Mogztter** [comment]
I will move forward and add `env-web-pdf` and `env=web-pdf` as they can be useful.

**Comment by Mogztter** [comment]
`env-web-pdf` and `env=web-pdf` attributes were added in: https://github.com/Mogztter/asciidoctor-web-pdf/commit/fa35c36f3525c4b43c2519b8c002e90f4b5ad343

Please open a new issue if this addition is not enough.

**Comment by mojavelinux** [comment]
This isn't quite the right use of env. The env is meant to indicate which environment Asciidoctor is being integrated into. For example, if I were to use this converter in Antora, then the env should be Antora, not web-pdf. The "web-pdf" tells me what the converter is, not the environment. So now we have a conflict.

I'm not sure I understand why backend=web-pdf and basebackend=html isn't enough. That tells you that the target format is PDF (specifically generated using the web stack) and that underneath it is using the HTML format.

I'm open to exploring new attributes if that doesn't sufficiently communicate the method of conversion.

**Comment by mojavelinux** [comment]
* backend=web-pdf (I understand if you want this to be pdf instead)
* basebackend=html
* outfilesuffix=.pdf
* filetype=pdf

We could consider a new attribute to qualify the variation of the backend. Perhaps one of these:

* converter=web-pdf (where backend=pdf; a converter satisfies a request for a backend)
* pdf-generator=web (as opposed to prawn)

Regardless, this converter should not set "env".

**Comment by Mogztter** [comment]
> This isn't quite the right use of env. The env attribute is meant to indicate which environment Asciidoctor is being integrated into. For example, if I were to use this converter in Antora, then the env should be antora, not web-pdf. The "web-pdf" value tells me what the converter is, not the environment. So now we have a conflict.

I see... 🤔 

> I'm not sure I understand why backend=web-pdf and basebackend=html isn't enough. That tells you that the target format is PDF (specifically generated using the web stack) and that underneath it is using the HTML format.

The value `web-pdf` is consistent with what we have today.
For instance, the backend for the reveal.js converter is `revealjs`. If we follow the same rule, then the value of the backend attribute for the Google Slides converter should be `googleslides` (or `google-slides`).
But, unless I'm mistaken, the value should be `slides` or `presentation`?
Otherwise, the value of the backend attribute is really the name of the converter.

Recently, I was asked to hide chunks of text when producing slides.
My recommendation was to avoid using "backend" because it's too specific.
If I have more than one converter producing slides/presentations then I will need to do something like:

```
ifndef::backend-revealjs+backend-googleslides[Not producing a reveal.js or Google Slides presentation.]
```

What is the recommendation in this case?

> * backend=web-pdf (I understand if you want this to be pdf instead)
> * basebackend=html
> * outfilesuffix=.pdf
> * filetype=pdf
>
> We could consider a new attribute to qualify the variation of the backend (in the case backend=pdf). Perhaps one of these:
> * converter=web-pdf (remember that a converter satisfies a request for a backend)
> * pdf-generator=web (as opposed to prawn)

I like this proposal because we provide exhaustive information.

**Comment by mojavelinux** [comment]
> unless I'm mistaken, the value should be slides or presentation? Otherwise, the value of the backend attribute is really the name of the converter.

Correct, that's really the idea behind it. Which is why I like better the idea of the converter setting an additional attribute to self-identify.

**Comment by mojavelinux** [comment]
> converter=web-pdf 

This could even be `backend-converter=web-pdf` (a readonly attribute)

**Comment by benoitrolland** [comment]
As a user, I need to make the difference betwen both the engine used and the format generated, 
the current value of  the "backend" attribute doesn't allow me to make the difference between html5 generated by asciidoctor and pdf generated by  asciidoctor-web-pdf, because in both case the value is "html5". 
My asciidoc files thus depends on external attributes (passed from the command-line)
I would expect the backend value to be one of asciidoctor, asciidoctor-pdf, asciidoctor-web-pdf, asciidoctor-fopub and another attribute like "output-format"

**Comment by mojavelinux** [comment]
I don't think resolution is obvious, otherwise we would have long since wrapped this up.

The base backend is html, but the converter goes on to subsequently convert that output to yet another format (PDF). So the basebackend has to be set at a minimum. (If you change the basebackend from html, certain features stop working, and that's not conducive for users). It's debatable whether the backend should be changed from html5 to something else too. If you do that, you lose some information tools may rely on. After all, the converter for the html5 backend **is** being used.

I put forth my suggestions in https://github.com/Mogztter/asciidoctor-web-pdf/issues/129#issuecomment-683349998. Another idea would be `backend-variant=web-pdf`. We need something to express that it is the html5 backend, but it is being used for a different purpose. (As I mentioned, this is not the right use case for `env`).

I also just remembered we had been discussing something like this in https://github.com/asciidoctor/asciidoctor/issues/3116.

> Another attribute like "output-format" would tell if the output will be epub3, html or pdf.

That's what 'filetype' is.

**Comment by benoitrolland** [comment]
ok thanks, so if `:backend:` is `html5` and `:filetype:` is pdf I can infer that the engine is `asciidoctor-web-pdf` ;-)

**Comment by mojavelinux** [comment]
That's not quite what I'm saying. If you look at [my proposal above](https://github.com/Mogztter/asciidoctor-web-pdf/issues/129#issuecomment-683349998) and Guillaume's reply that followed it, I think we agree it should be `backend=web-pdf` `basebackend=html` `filetype=pdf`. However, if we decided for some reason not to set backend, then you would look for `backend-converter` to have the value web-pdf. But that's not an attribute we have ever had in Asciidoctor, so I'm less comfortable with it.

**Comment by mojavelinux** [comment]
> so if :backend: is html5 and :filetype: is pdf,

Let's just say you wanted to take this approach, except you check `basebackend=html` and `filetype=pdf`. Then you would only know that Asciidoctor is producing a PDF with an HTML-based PDF converter. Asciidoctor PDF fits that criteria, so it won't tell you you are using Asciidoctor Web PDF. That's an argument for why `backend` should be set to `web-pdf`. Because only then does it uniquely identify which converter is being used. However, it violates the design of the backend, which is that you should not care which converter is registered for it. Rather, you just care that you are producing pdf. That makes me think that, if you absolutely must know which converter it is, then we really should have a `converter` attribute that provides this information.

In light of that, I'd like to revise my proposal:

* backend=pdf
* basebackend=html
* filetype=pdf
* converter=web-pdf
* pdf-generator=browser (or puppeteer)

Like I said, this decision isn't easy.

**Comment by benoitrolland** [comment]
In my specific case I need to know what converter is used when features are differently implemented.
This is the case for the [index], 
I need to print an alternative when not available like in web.pdf .
